graphql_compiler/compiler/emit_cypher.py

```diff
@@ -127,10 +159,120 @@ def _emit_with_clause_components(cypher_steps):
+            result.append(u',')
+        result.append(u'\n  %(collect_call)s AS %(collected_name)s' %
+                      {'collect_call': collect_call, 'collected_name': vertex_names[collect_call]})
+    return result
+
+
+def _emit_fold_scope(cypher_query):
+    """Return a Cypher query pattern match expression for each fold scope in cypher_query.
+
+    Consider the following CYPHER query, corresponding to test_input_data.multiple_folds()
+
+    MATCH (Animal___1:Animal)
+    OPTIONAL MATCH (Animal___1)-[:Animal_ParentOf]->(Animal__out_Animal_ParentOf___1:Animal)
+    WITH
+      Animal___1 AS Animal___1,
+      collect(Animal__out_Animal_ParentOf___1) AS collected_Animal__out_Animal_ParentOf___1
+    OPTIONAL MATCH (Animal___1)<-[:Animal_ParentOf]-(Animal__in_Animal_ParentOf___1:Animal)
+    WITH
+      Animal___1 AS Animal___1,
+      collected_Animal__out_Animal_ParentOf___1 AS
+        collected_Animal__out_Animal_ParentOf___1,
+      collect(Animal__in_Animal_ParentOf___1) AS
+        collected_Animal__in_Animal_ParentOf___1
+    RETURN
+      Animal___1.name AS `animal_name`,
+      [x IN collected_Animal__out_Animal_ParentOf___1 | x.name] AS `child_names_list`,
+      [x IN collected_Animal__out_Animal_ParentOf___1 | x.uuid] AS `child_uuids_list`,
+      [x IN collected_Animal__in_Animal_ParentOf___1 | x.name] AS `parent_names_list`,
+      [x IN collected_Animal__in_Animal_ParentOf___1 | x.uuid] AS `parent_uuids_list`
+
+    For each fold scope, the structure generally works as follows:
+
+    First, traverse to the vertex just outside the fold scope. Each traversal is mandatory here, so
+    each traversal here is represented by a MATCH clause. Corresponds to:
+        MATCH (Animal___1:Animal)
+
+    Then, for each vertex at/ inside the fold scope, add a corresponding OPTIONAL MATCH clause.
+    This traversal is marked OPTIONAL MATCH and not MATCH since we want the result to be an empty
+    list and still show up in the final result even if nothing matches the traversal. Corresponds
+    to:
+        OPTIONAL MATCH (Animal___1)-[:Animal_ParentOf]->(Animal__out_Animal_ParentOf___1:Animal)
+
+    After traversing to the innermost scope (at which point all the OPTIONAL MATCH clauses are
+    generated, create a WITH clause. For each previously-expanded vertex V not in a fold scope, pass
+    it through the WITH clause as itself (i.e. `V AS V`). For each vertex at or inside a
+    previously-seen fold scope, return the name given to the list. For vertices at or inside the
+    current fold scope, collect the vertices in the WITH clause as well, which materializes the list
+    and guarantees that the order of elements within the output of a @fold is stable within each
+    result set. Corresponds to:
+        WITH
+          Animal___1 AS Animal___1,
+          collect(Animal__out_Animal_ParentOf___1) AS collected_Animal__out_Animal_ParentOf___1
+
+    This process is repeated for other fold scopes in the query. Corresponds to:
```

Comment:
Nice, this helps a lot. For bonus points, use a numbered list instead of bullet points, and then tie the comments in the code into the numbered list by making them use matching numbering. A more hard-core approach (which might be overkill here, but just mentioning it for inspiration) exists in the `_compile_vertex_ast()` function in `compiler_frontend.py`.

Reply:
I like the numbering idea!